File: python/sorting/mergesort.py
"""
Mergesort
Time Complexity: O(n log n)
Space Complexity: O(n)

Description:
A stable, divide-and-conquer sorting algorithm. It divides the array into halves,
recursively sorts them, and then merges the sorted halves.

Use when:
- Stable sorting is required.
- Predictable performance is needed.
"""

def mergesort(arr):
    if len(arr) <= 1:
        return arr
    
    mid = len(arr) // 2
    left = mergesort(arr[:mid])
    right = mergesort(arr[mid:])
    
    return merge(left, right)

def merge(left, right):
    result = []
    l = r = 0
    
    while l < len(left) and r < len(right):
        if left[l] <= right[r]:
            result.append(left[l])
            l += 1
        else:
            result.append(right[r])
            r += 1
            
    result.extend(left[l:])
    result.extend(right[r:])
    return result

if __name__ == "__main__":
    test_arr = [38, 27, 43, 3, 9, 82, 10]
    sorted_arr = mergesort(test_arr)
    assert sorted_arr == [3, 9, 10, 27, 38, 43, 82]
    print("Python Mergesort: All tests passed!")

File: typescript/src/sorting/mergesort.ts
/**
 * Mergesort
 * 
 * A stable, comparison-based, divide-and-conquer sorting algorithm.
 * 
 * Time Complexity: O(n log n) (consistent performance)
 * Space Complexity: O(n) (requires extra space for merging)
 * 
 * Use when:
 * - Stable sorting is required.
 * - Predictable O(n log n) performance is needed.
 * - Sorting linked lists (can be done in O(1) extra space).
 * 
 * Avoid when:
 * - Memory is highly constrained (use Quicksort or Heapsort).
 */
export function mergesort<T>(arr: T[]): T[] {
  if (arr.length <= 1) return arr;

  const mid = Math.floor(arr.length / 2);
  const left = mergesort(arr.slice(0, mid));
  const right = mergesort(arr.slice(mid));

  return merge(left, right);
}

function merge<T>(left: T[], right: T[]): T[] {
  const result: T[] = [];
  let l = 0;
  let r = 0;

  while (l < left.length && r < right.length) {
    if (left[l] <= right[r]) {
      result.push(left[l]);
      l++;
    } else {
      result.push(right[r]);
      r++;
    }
  }

  return [...result, ...left.slice(l), ...right.slice(r)];
}
